### PR TITLE
Support AOT mode for Windows

### PR DESCRIPTION
**Pull request by stuartmorgan**
- Adds a way to provide an AOT library to the C API.
- Adds app.so to the information provided by DartProject.
- Fixes the engine to only do static snapshot linking for Windows in
  debug mode, not all modes, so that the provided library is used.

Engine side of https://github.com/flutter/flutter/issues/38477

### CODE REVIEW

File: shell/platform/windows/flutter_windows.cc [hunk 1]
```diff
@@ -29,6 +30,33 @@
...
+        << std::endl;
+    return nullptr;
+  }
+  if (!std::filesystem::exists(aot_data_path)) {
+    std::cerr << "Can't load AOT data from " << aot_data_path.u8string()
+              << "; no such file." << std::endl;
+    return nullptr;
+  }
+  std::string path_string = aot_data_path.u8string();
+  FlutterEngineAOTDataSource source;
```
Comment (chinmaygarde) [inline shell/platform/windows/flutter_windows.cc:48]: Nit: Zero out C structs please. `= {}` here and `= nullptr` later when specifying the `data` variable. The struct thing is important for all structs specified by the Embedder API (I realize it doesn't matter for this particular struct which is a tagged union but it is just a good habit to get into IMO).

File: shell/platform/windows/flutter_windows.cc [hunk 2]
```diff
@@ -120,10 +153,22 @@ static std::unique_ptr<FlutterDesktopEngineState> RunFlutterEngine(
...
+    if (!aot_library_path.empty()) {
+      aot_library_path =
+          std::filesystem::path(executable_location) / aot_library_path;
+    }
   }
   std::string assets_path_string = assets_path.u8string();
   std::string icu_path_string = icu_path.u8string();
 
+  if (FlutterEngineRunsAOTCompiledDartCode()) {
+    state->aot_data = std::move(LoadAotData(aot_library_path));
```
Comment (chinmaygarde) [inline shell/platform/windows/flutter_windows.cc:165]: Nit: The `std::move` is redundant for an rvalue.

File: shell/platform/windows/window_state.h [hunk 3]
```diff
@@ -39,13 +39,24 @@ struct FlutterDesktopView {
   flutter::Win32FlutterWindow* window;
 };
 
+struct AotDataDeleter {
+  void operator()(FlutterEngineAOTData aot_data) {
+    FlutterEngineCollectAOTData(aot_data);
```
Comment (chinmaygarde) [inline shell/platform/windows/window_state.h:44]: Nit: Not sure if null aot_data's can be collected. Maybe a paranoid check here?

File: shell/platform/windows/public/flutter_windows.h [hunk 4]
```diff
@@ -8,12 +8,11 @@
 #include <stddef.h>
 #include <stdint.h>
 
+#include "Windows.h"
```
Comment (chinmaygarde) [inline shell/platform/windows/public/flutter_windows.h:11]: Is this a system header? If so, it needs to be included with <>.

Comment (stuartmorgan) [reply shell/platform/windows/public/flutter_windows.h]: Oops, somehow I've never noticed that was wrong while updating headers in this file.

File: shell/platform/windows/window_state.h (hunk 3 above)
Comment (stuartmorgan) [reply shell/platform/windows/window_state.h:44]: FlutterEngineCollectAOTData has an internal null check.

File: shell/platform/windows/flutter_windows.cc (hunk 1 above)
Comment (stuartmorgan) [reply shell/platform/windows/flutter_windows.cc]: Good catch, thanks.

File: shell/platform/windows/flutter_windows.cc (hunk 2 above)
Comment (stuartmorgan) [reply shell/platform/windows/flutter_windows.cc]: Done.

**Comment by stuartmorgan** [comment]
Mac Android Debug Engine and Mac Web Engine were both green on re-run.

**Pull request merged**